proof/refine/ARM/Finalise_R.thy

```diff
@@ -4362,62 +4489,388 @@ global_interpretation delete_one
+                         sc_relation (f (\<lambda>_. new) sc) n (f' sc')"
+  assumes y: "\<And>sc. sc_replies sc = sc_replies (f (\<lambda>_. new) sc)"
+  assumes z: "\<And>sc'. objBits sc' = objBits (f' sc')"
+  assumes a:  "scReply sc' = scReply (f' sc')"
+  shows      "corres dc (sc_at scPtr)
+                        (ko_at' sc' scPtr)
+                    (set_sc_obj_ref f scPtr new) (setSchedContext scPtr (f' sc'))"
+  unfolding update_sched_context_def
+  apply (rule corres_guard_imp)
+    apply (rule corres_symb_exec_l'[where Q'="\<lambda>r s. ko_at r scPtr s \<and> (\<exists>n. is_sc_obj n r)",
+                                    where P="\<lambda>s. obj_at \<top> scPtr s"])
+      apply (rule corres_guard_imp[where P'=R and Q'=R for R])
+        apply (rule_tac F="(\<exists>n. is_sc_obj n obj)" in corres_gen_asm)
+        apply (case_tac obj; simp add: is_sc_obj_def)
+        apply (rule setSchedContext_no_stack_update_corres[where f'=f'])
+           apply (clarsimp simp: x obj_at_def intro!: y z a)+
+     apply (fastforce simp: exs_valid_def get_object_def in_monad)
+    apply (wpsimp wp: get_object_wp)
+   apply (fastforce simp: obj_at_def)
+  apply simp
+  done
+
+lemma schedContextUnbindTCB_corres:
+  "corres dc (einvs and sc_tcb_sc_at bound sc_ptr)
+             (invs' and obj_at' (\<lambda>sc. bound (scTCB sc)) sc_ptr)
+          (sched_context_unbind_tcb sc_ptr) (schedContextUnbindTCB sc_ptr)"
+  apply (clarsimp simp: sched_context_unbind_tcb_def schedContextUnbindTCB_def
+                        sym_refs_asrt_def)
+  apply add_sym_refs
+  apply (rule corres_stateAssert_implied[where P'=\<top>, simplified])
+  apply (rule corres_guard_imp)
+     apply (rule corres_split_new[OF get_sc_corres])
+       apply (rule corres_assert_opt_assume_l)
+       apply (rule corres_assert_assume_r)
+       apply (prop_tac "scTCB sca = sc_tcb sc"; clarsimp)
+        apply (clarsimp simp: sc_relation_def)
+       apply (rule corres_split_new[OF gct_corres])
+         apply (rule corres_split_new[OF corres_when], clarsimp simp: sc_relation_def)
+            apply (rule rescheduleRequired_corres)
+           apply (rule corres_split_new[OF tcbSchedDequeue_corres])
+             apply (rule corres_split_new[OF tcb_release_remove_corres])
+               apply (rule corres_split_new[OF set_tcb_obj_ref_corresT])
+                    apply (clarsimp simp: tcb_relation_def)
+                   apply (rule ball_tcb_cap_casesI; simp)
+                  apply (clarsimp simp: tcb_cte_cases_def)
+                 apply (rule_tac sc'=sca in set_sc_obj_ref_no_stack_update_corres)
+                    apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def)+
+                apply wpsimp+
+       apply (case_tac sca; clarsimp)
+       apply (wpfix add: sched_context.sel)
+       apply simp
+      apply wpsimp+
+    apply (frule invs_valid_objs)
+    apply (auto simp: sc_at_pred_n_def obj_at_def is_obj_defs valid_obj_def
```

Comment:
In that discussion on `by`, I know Matt said he was dead against `auto (blah)[1]`. Can this be removed, or maybe we do want to go with `subgoal by auto (blah)` to help with some sort of CI check?

Just out of curiosity, I've seen a lot of `frule`s followed by big `fastforce` or `auto` blocks. Is there something preventing `auto` or `fastforce` from using `invs_valid_objs` as a `dest` rule in these cases?

Reply:
In this case I can just change the `auto` to a `fastforce`. I'm not sure why I didn't do that originally.

I think generally we use an `frule` like that when there's an assumption that needs to be used multiple times. Here there's multiple facts that we need from `invs'` (we need to show `pspace_aligned` and `pspace_distinct`, and the `valid_objs` is required for getting to a `valid_obj`).

Reply:
Ah, that makes sense. Cool